alchemy/src/auth.ts

```diff
@@ -0,0 +1,326 @@
+      credentials,
+    );
+  };
+
+  /**
+   * Deletes the credentials file.
+   * @param props The profile and provider of the credentials.
+   */
+  export const del = async (props: Props) => {
+    await fs.unlink(Path.credentialsFile(props.provider, props.profile));
+  };
+
+  /**
+   * Gets refreshed OAuth credentials.
+   * Uses `singleFlight` and `Lock` to avoid making multiple concurrent requests to refresh credentials.
+   * @param props The properties of the credentials.
+   * @param refresh The function to refresh the credentials.
+   * @returns The refreshed credentials.
+   */
+  export const getRefreshed = singleFlight(
+    async (
+      props: Props,
+      refresh: (credentials: Credentials.OAuth) => Promise<Credentials.OAuth>,
+    ): Promise<Credentials> => {
+      const credentials = await Credentials.get(props);
+      if (!credentials) {
+        throw new Error(
+          `Credentials for provider "${props.provider}" not found in profile "${props.profile}"`,
+        );
+      }
+      if (!Credentials.isOAuthExpired(credentials)) {
+        return credentials;
+      }
+      const key = `${props.provider}-${props.profile}`;
+      if (await Lock.acquire(key)) {
+        try {
+          const refreshed = await refresh(credentials);
+          await Credentials.set(props, refreshed);
+          return refreshed;
+        } finally {
+          await Lock.release(key);
+        }
+      }
+      await Lock.wait(key);
+      return await Credentials.getRefreshed(props, refresh);
+    },
+    (props) => `${props.provider}:${props.profile}`,
+  );
+
+  /**
+   * Returns true if the given credentials are OAuth and expired.
+   */
+  export const isOAuthExpired = (
+    credentials: Credentials,
+    tolerance = 1000 * 10,
+  ): credentials is Credentials.OAuth => {
+    return (
+      credentials.type === "oauth" &&
+      credentials.expires < Date.now() + tolerance
+    );
+  };
+}
+
+namespace FS {
+  export const readJSON = async <T>(path: string): Promise<T | undefined> => {
+    try {
+      const data = await fs.readFile(path, "utf-8");
+      return JSON.parse(data) as T;
+    } catch {
+      return undefined;
+    }
+  };
+
+  export const writeJSON = async <T>(name: string, data: T) => {
+    await fs.mkdir(path.dirname(name), { recursive: true });
+    await fs.writeFile(name, JSON.stringify(data, null, 2), { mode: 0o600 });
+  };
+}
+
+namespace Lock {
+  const STALE = 1000 * 10;
+
+  interface File {
+    name: string;
+    pid: number;
+    timestamp: number;
+  }
+
+  /**
+   * Acquires a lock on a resource.
+   *
+   * WARNING: This function uses async io, so it is not thread-safe within a single process.
+   * Use a `singleFlight` or mutex to ensure thread-safety.
+   *
+   * @param name The name of the resource to lock.
+   * @returns True if the lock was acquired, false if it was already held by another process.
+   */
+  export const acquire = async (name: string): Promise<boolean> => {
```

Comment:
Why don't we lock internally?

Reply:
I am having a hard time finding the lock implementation?

I have used https://www.npmjs.com/package/proper-lockfile and it's pretty useful if you need multi-process locking. 

Do we need to worry about turborepos trying to access profiles at the same time?

Reply:
Here's how it's used:

https://github.com/sam-goodwin/alchemy/pull/956/files#diff-3f8f32dccc166808066dbb4b172417239d789686509c3a7497327b359691567fR194-R208

Here's how it's implemented:

https://github.com/sam-goodwin/alchemy/blob/72fbc41cbd87cf19657863d6c5b759a602e485c7/alchemy/src/util/lock.ts

Happy to share my test scripts for it - I would've checked them in but frankly wasn't sure where to put them because it involved transpiling and spawning to verify it works in Node.